cvmfs/cache_posix.cc

```diff
@@ -310,14 +335,44 @@ void *PosixCacheManager::DoSaveState() {
 
 int PosixCacheManager::DoRestoreState(void *data) {
   assert(data);
+  if (do_refcount_) {
+    SavedState *state = reinterpret_cast<SavedState *>(data);
+    if (state->magic_number == 123) {
+      LogCvmfs(kLogCache, kLogDebug, "Restoring refcount cache manager from "
+                                    "refcounted posix cache manager");
+
+      fd_mgr->AssignFrom(state->fd_mgr.weak_ref());
+    } else {
+      LogCvmfs(kLogCache, kLogDebug, "Restoring refcount cache manager from "
+                                    "non-refcounted posix cache manager");
+    }
+    return -1;
+  }
+
   char *c = reinterpret_cast<char *>(data);
-  assert(*c == '\0');
+  assert(*c == '\0' || *c == 123);
+  if (*c == 123) {
+    SavedState *state = reinterpret_cast<SavedState *>(data);
+    LogCvmfs(kLogCache, kLogDebug, "Restoring non-refcount cache manager from "
+                                    "refcounted posix cache manager - this "
+                                    " is not possible, keep refcounting.");
+    fd_mgr->AssignFrom(state->fd_mgr.weak_ref());
+    do_refcount_ = true;
+  }
   return -1;
 }
 
 
 bool PosixCacheManager::DoFreeState(void *data) {
-  free(data);
+  assert(data);
+  SavedState *state = reinterpret_cast<SavedState *>(data);
+  if (state->magic_number == 123) {
+    delete state;
+  } else {
+    // this should be the dummy SavedState
+    // of the regular posix cache manager
+    free(data);
+  }
```

Comment:
In both cases we need to `free(data)`
```suggestion
  } 
  // If not refcounted, the state is the dummy SavedState
  // of the regular posix cache manager
  free(data);
```